File: src/main/scala/fortress/solverinterface/Z3ApiSolver.java
package fortress.solverinterface;

import java.io.*;

import fortress.msfol.*;
import fortress.outputs.*;
import fortress.util.Pair;
import fortress.util.StopWatch;
import fortress.util.Errors;

import fortress.modelfind.*;
import fortress.solverinterface.*;
import fortress.interpretation.*;

import com.microsoft.z3.*;

public class Z3ApiSolver extends SolverTemplate {
    private Model lastModel = null;
    private Context context = null;
    private Solver solver = null;
    
    @Override
    public boolean canAttemptSolving(Theory theory) {
        return true;
    }
    
    @Override
    protected void convertTheory(Theory theory, Writer log) { 
        var pair = new TheoryToZ3Java(theory).convert();
        context = pair._1();
        solver = pair._2();
    }
    
    @Override
    protected void updateTimeout(int remainingMillis) {
        Errors.assertion(null != context);
        Errors.assertion(null != solver);
        
        Params params = context.mkParams();
        params.add("timeout", remainingMillis);
        solver.setParameters(params);
    }
    
    @Override
    protected ModelFinderResult runSolver(Writer log) throws IOException {
        Errors.assertion(null != context);
        Errors.assertion(null != solver);
        
        Status status = solver.check();
        lastModel = null;
        switch(status) {
            case UNKNOWN:
                // TODO timeout errors
                log.write("UKNOWN (" + solver.getReasonUnknown() + ").\n");
                if(solver.getReasonUnknown().equals("timeout")
                        || solver.getReasonUnknown().equals("canceled")) {
                    return ModelFinderResult.Timeout();
                }
                return ModelFinderResult.Unknown();
                // break;
            case SATISFIABLE:
                lastModel = solver.getModel();
                log.write("SAT.\n");
                return ModelFinderResult.Sat();
                // break;
            case UNSATISFIABLE:
                log.write("UNSAT.\n");
                return ModelFinderResult.Unsat();
                // break;
            default:
                throw new RuntimeException("Unexpected solver result " + status.toString());
        }
    }

    public Interpretation getInstance(Theory theory) {
        Errors.assertion(null != lastModel, "There is no current instance");
        return new Z3ApiInterpretation(lastModel, theory.signature());
    }
}

File: src/main/scala/fortress/solverinterface/Z3ApiSolver.scala
package fortress.solverinterface

import fortress.msfol._
import fortress.util._

import fortress.modelfind._
import fortress.solverinterface._
import fortress.interpretation._

import com.microsoft.z3.{
    Model => Z3Model,
    Context => Z3Context,
    Solver => Z3Solver,
    Status => Z3Status,
    Params => Z3Params
}

class Z3ApiSolver extends SolverTemplate {
    private var lastModel: Option[Z3Model] = None
    private var context: Option[Z3Context] = None
    private var solver: Option[Z3Solver] = None
    private var converter: Option[TheoryToZ3_StringParse] = None

    override protected def convertTheory(theory: Theory): Unit = {
        converter = Some(new TheoryToZ3_StringParse(theory))
        val pair: (Z3Context, Z3Solver) = converter.get.convert
        context = Some(pair._1)
        solver = Some(pair._2)
    }

    override def addAxiom(axiom: Term, timeoutMillis: Milliseconds): ModelFinderResult = {
        solver.get.push()
        solver.get.add(converter.get.convertAxiom(axiom))

        updateTimeout(timeoutMillis)
        runSolver()
    }

    override protected def updateTimeout(remainingMillis: Milliseconds): Unit = {
        Errors.assertion(context.nonEmpty)
        Errors.assertion(solver.nonEmpty)

        val params: Z3Params = context.get.mkParams()
        params.add("timeout", remainingMillis.value)
        solver.get.setParameters(params)
    }

    @throws(classOf[java.io.IOException])
    override protected def runSolver(): ModelFinderResult = {
        Errors.assertion(context.nonEmpty)
        Errors.assertion(solver.nonEmpty)

        val status: Z3Status = solver.get.check()
        lastModel = None
        status match {
            case Z3Status.UNKNOWN => {
                // TODO timeout errors
                // log.write("UNKNOWN (" + solver.get.getReasonUnknown() + ").\n")
                if(solver.get.getReasonUnknown() == "timeout"
                        || solver.get.getReasonUnknown == "canceled") {
                    return ModelFinderResult.Timeout
                }
                return ModelFinderResult.Unknown
            }
            case Z3Status.SATISFIABLE => {
                lastModel = Some(solver.get.getModel())
                // log.write("SAT.\n")
                return ModelFinderResult.Sat
            }
            case Z3Status.UNSATISFIABLE => {
                // log.write("UNSAT.\n")
                return ModelFinderResult.Unsat
            }
            case _ => throw new java.lang.RuntimeException("Unexpected solver result " + status.toString)
        }
    }

    def getInstance(theory: Theory): Interpretation = {
        Errors.assertion(lastModel.nonEmpty, "There is no current instance")
        return new Z3ApiInterpretation(lastModel.get, theory.signature, converter.get, context.get)
    }
}
